eclair-core/src/main/scala/fr/acinq/eclair/router/Router.scala

```diff
@@ -393,17 +409,52 @@ object Router {
     def cltvExpiryDelta: CltvExpiryDelta
   }
 
+  // @formatter:off
+  /** source of the channel routing parameters */
+  // TODO: rename this channel params or routing params or ???
+  sealed trait ChannelSource {
+    def cltvExpiryDelta: CltvExpiryDelta
+    def relayFees: Relayer.RelayFees
+    def fee(amount: MilliSatoshi): MilliSatoshi = nodeFee(relayFees, amount)
+    def htlcMinimum: MilliSatoshi
+    def htlcMaximum_opt: Option[MilliSatoshi]
+  }
+  object ChannelSource {
+    /** we learnt about this channel from a channel_update */
+    case class Announcement(channelUpdate: ChannelUpdate) extends ChannelSource {
+      lazy val cltvExpiryDelta: CltvExpiryDelta = channelUpdate.cltvExpiryDelta
+      lazy val relayFees: Relayer.RelayFees = RelayFees(channelUpdate.feeBaseMsat, channelUpdate.feeProportionalMillionths)
+      lazy val htlcMinimum: MilliSatoshi = channelUpdate.htlcMinimumMsat
+      lazy val htlcMaximum_opt: Option[MilliSatoshi] = channelUpdate.htlcMaximumMsat
+    }
+    /** we learnt about this channel from hints in an invoice */
+    case class Hint(extraHop: Bolt11Invoice.ExtraHop, htlcMaximum: MilliSatoshi) extends ChannelSource {
+      lazy val cltvExpiryDelta: CltvExpiryDelta = extraHop.cltvExpiryDelta
+      lazy val relayFees: Relayer.RelayFees = RelayFees(extraHop.feeBase, extraHop.feeProportionalMillionths)
+      lazy val htlcMinimum: MilliSatoshi = 0 msat
+      lazy val htlcMaximum_opt: Option[MilliSatoshi] = Some(htlcMaximum)
+    }
+
+    def areSame(source1: ChannelSource, source2: ChannelSource): Boolean =
+      source1.relayFees == source2.relayFees &&
+        source1.cltvExpiryDelta == source2.cltvExpiryDelta &&
+        source1.htlcMinimum == source2.htlcMinimum // TODO: not comparing htlcMax otherwise network/invoice sources may never match
+  }
+  // @formatter:on
+
   /**
    * A directed hop between two connected nodes using a specific channel.
    *
-   * @param nodeId     id of the start node.
-   * @param nextNodeId id of the end node.
-   * @param lastUpdate last update of the channel used for the hop.
+   * @param nodeId         id of the start node.
+   * @param nextNodeId     id of the end node.
+   * @param shortChannelId scid that will be used to build the payment onion.
+   * @param source         source for the channel parameters.
    */
-  case class ChannelHop(nodeId: PublicKey, nextNodeId: PublicKey, lastUpdate: ChannelUpdate) extends Hop {
-    override lazy val cltvExpiryDelta: CltvExpiryDelta = lastUpdate.cltvExpiryDelta
-
-    override def fee(amount: MilliSatoshi): MilliSatoshi = nodeFee(lastUpdate, amount)
+  case class ChannelHop(shortChannelId: ShortChannelId, nodeId: PublicKey, nextNodeId: PublicKey, source: ChannelSource) extends Hop {
+    // @formatter:off
+    override def cltvExpiryDelta: CltvExpiryDelta = source.cltvExpiryDelta
+    override def fee(amount: MilliSatoshi): MilliSatoshi = source.fee(amount)
```

Comment:
I went back and forth several times between this model and the following, where `ChannelSource` is a simple data container:

````scala
sealed trait ChannelSource
object {
  case class Announcement(channelUpdate: ChannelUpdate) extends ChannelSource
  case class Hint(extraHop: Bolt11Invoice.ExtraHop, htlcMaximum: MilliSatoshi) extends ChannelSource
}

case class ChannelHop(shortChannelId: ShortChannelId, nodeId: PublicKey, nextNodeId: PublicKey, source: ChannelSource) extends Hop {
    override def cltvExpiryDelta: CltvExpiryDelta = source match {
      case s: ChannelSource.Announcement => s.channelUpdate.cltvExpiryDelta
      case s: ChannelSource.Hint => s.extraHop.cltvExpiryDelta
    }
    // ...
````

The model I retained more cleanly defines what a `ChannelSource` is and better separate concerns, but it's more verbose and calling `channelHop.source.htlcMinimum`  is a bit ugly. :-/ I think it's because we have to conflicting abstractions: Channel/Node and Annoucement/Hint

Reply:
> I think the issue around naming is because you're trying to do two different things at the same time:
>
>- factorize channel relay parameters
>- label them by different sources

Exactly. Also, there is kind of a competition between `ChannelHop` and `GraphEdge`, which are very similar and both take a `source` as argument. For `GraphEdge` it makes more sense to focus on the source because we may need to update it, whereas for `ChannelHop` we mainly want the parameters to build the onion. Maybe we should revisit after we have experimented with the new edge model suggested by @thomash-acinq. The idea is to index edges by (node1, node2) instead of channel id. That means each edge would contain a set of channels, and we would somehow combine all those into one.

> What do you think about the following model, which explicitly separates the two purposes?

I would prefer keeping the original data, and I don't like the use of `object`s as enums .

I went in b673910da with a simple rename `ChannelSource`->`ChannelRelayParams` which has the advantage of not creating potential issues with further changes that I have in the pipe:
- it now focuses on the "params" aspect rather than the "source"
- the type name still makes the source obvious, : `ChannelRelayParams.FromAnnouncement`/`ChannelRelayParams.FromHint`
- `channelHop.params.htlcMinimum` looks better than `channelHop.source.htlcMinimum`